actors/cache/actors_cache.go

```diff
@@ -134,16 +133,51 @@ func (a *ActorsCache) GetActorCode(add address.Address, key filTypes.TipSetKey,
 		return "", fmt.Errorf(" %w : %s is flagged as bad", ErrBadAddress, addStr)
 	}
 
-	if !onChainOnly {
-		actorCode, err := a.offChainCache.GetActorCode(add, key, onChainOnly, canonical)
-		if err == nil {
-			return actorCode, nil
+	getActorCodeFn := func() (string, error) {
+		var actorCode string
+		var err error
+		if canonical {
+			// Try canonical cache
+			actorCode, err = a.offChainCache.GetActorCode(add, key, onChainOnly, true)
+			if err == nil {
+				return actorCode, nil
+			}
+			a.logger.Debugf("[ActorsCache] - Unable to retrieve actor code from canonical offchain cache for address %s. Trying on-chain cache", addStr)
+			// Try onchain
+			actorCode, err = a.onChainCache.GetActorCode(add, key, onChainOnly, true)
+			if err == nil {
+				return actorCode, nil
+			}
+			a.logger.Debugf("[ActorsCache] - Unable to retrieve actor code from onchain cache for address %s. Trying latest cache", addStr)
+
+			// Try latest cache
+			actorCode, err = a.offChainCache.GetActorCode(add, key, onChainOnly, false)
+			if err == nil {
+				return actorCode, nil
+			}
```

Comment:
I think this one should never happen. Canonical should use canonical cache or onchain. Latest may hold invalid relationship.